pkg/registry/registry.go

```diff
@@ -54,18 +55,44 @@ func (v *v) v2(resp http.ResponseWriter, req *http.Request) *regError {
 	return nil
 }
 
-func (v *v) root(resp http.ResponseWriter, req *http.Request) {
-	if rerr := v.v2(resp, req); rerr != nil {
-		log.Printf("%s %s %d %s %s", req.Method, req.URL, rerr.Status, rerr.Code, rerr.Message)
+func (r *registry) root(resp http.ResponseWriter, req *http.Request) {
+	if rerr := r.v2(resp, req); rerr != nil {
+		r.logf("%s %s %d %s %s", req.Method, req.URL, rerr.Status, rerr.Code, rerr.Message)
 		rerr.Write(resp)
 		return
 	}
-	log.Printf("%s %s", req.Method, req.URL)
+	r.logf("%s %s", req.Method, req.URL)
 }
 
-// New returns a handler which implements the docker registry protocol. It should be registered at the site root.
+func (r *registry) logf(f string, v ...interface{}) {
+	if r.log == nil {
+		log.Printf(f, v...)
+	} else {
+		r.log.Printf(f, v...)
+	}
+}
+
+// New returns a handler which implements the docker registry protocol.
+// It should be registered at the site root.
 func New() http.Handler {
-	v := v{
+	return NewWithOptions(nil)
+}
+
+// Options describes the available options
+// for creating the registry.
+type Options struct {
+	// Log is used to log requests.
+	// If nil, the global logger is used.
+	Log *log.Logger
+}
```

Comment:
Looks awesome - I think this is definitely a pretty reasonable thing to change and thank you for sending in the change.

They style in go is generally to define a option type - i.e.

type Option func(o *registry)

And then have New take variadic arguments - i.e.
New(o Option...)

and in the body do something like

opts := &registry{log:defaultLog}
for _, o := range options {
 o(opts)
}

And define the logger option like
func WithLogger(l log.Type) Option

so you can call New() to get default behaviour, New(WithLogger(log.SomethingComplicated)) to get more interesting behaviour.

Let me know if that makes sense - I'm happy to have the log be configurable, just want to make sure the option pattern is nice for the user. I'm also happy to make these changes if I'm asking for too much work from you.

Reply:
> They style in go is generally to define a option type - i.e.

Its another pattern for sure but I wouldn't say its generally better. The stdlib for example doesn't use it at all and even though later packages did, the new go-cloud package made by the Go team explicitly decided against using them in favour of optional structs.

See https://github.com/google/go-cloud/issues/908#issuecomment-445034143

Reply:
> I'm also happy to make these changes if I'm asking for too much work from you.

Feel free to take over.

Reply:
I think both approaches are reasonable. There are definitely some hard to solve problems with the functional options approach, but it looks so much nicer in some cases 😄 

We should keep in mind any other extension points e.g. https://github.com/google/go-containerregistry/issues/488

Reply:
I pushed a version of this in https://github.com/google/go-containerregistry/pull/537 - let me know what you think :D